File: cpp/86_partition_list.cpp
#include <bits/stdc++.h>
using namespace std;

/*Definition for singly-linked list.*/
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* partition(ListNode* head, int x) {

        if (head == nullptr)
            return head;

        ListNode less_head = ListNode(), greater_head = ListNode();
        ListNode* less = nullptr;
        ListNode* greater = nullptr;
        ListNode* cur = head;

        while (cur != nullptr) {
            if (cur->val < x) {
                if (less == nullptr) {
                    less_head.next = cur;
                    less = cur;
                } else {
                    less->next = cur;
                    less = cur;
                }
            } else {
                if (greater == nullptr) {
                    greater_head.next = cur;
                    greater = cur;
                } else {
                    greater->next = cur;
                    greater = cur;
                }
            }
            cur = cur->next;
        }

        if (less_head.next == nullptr) {
            return greater_head.next;
        } else {
            less->next = greater_head.next;
            if (greater != nullptr)
                greater->next = nullptr;
            return less_head.next;
        }

    }
};

File: go/86_partition_list.go
package main

func partition(head *ListNode, x int) *ListNode {

	if head == nil {
		return head
	}

	var smol_head *ListNode = nil
	var smol_cur *ListNode = nil
	var big_head *ListNode = nil
	var big_cur *ListNode = nil

	cur := head
	for cur != nil {
		if cur.Val < x {
			if smol_head == nil {
				smol_head = cur
				smol_cur = cur
			} else {
				smol_cur.Next = cur
				smol_cur = smol_cur.Next
			}
		} else {
			if big_head == nil {
				big_head = cur
				big_cur = cur
			} else {
				big_cur.Next = cur
				big_cur = big_cur.Next
			}
		}
		cur = cur.Next
	}

	if smol_head == nil {
		big_cur.Next = nil
		return big_head
	} else {
		smol_cur.Next = big_head
		if big_cur != nil {
			big_cur.Next = nil
		}
		return smol_head
	}
}
